File: apps/cli/src/commands/export-site.ts
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { Command } from "commander";
import { formatSuccess } from "../lib/colors";
import { output, outputError } from "../lib/output";
import { getServerUrl } from "../lib/store";

function escapeHtml(str: string): string {
    return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}

function markdownToHtml(md: string): string {
    // Minimal markdown-to-HTML for chunk content
    return md
        .replace(/```(\w*)\n([\s\S]*?)```/g, "<pre><code>$2</code></pre>")
        .replace(/^### (.+)$/gm, "<h3>$1</h3>")
        .replace(/^## (.+)$/gm, "<h2>$1</h2>")
        .replace(/^# (.+)$/gm, "<h1>$1</h1>")
        .replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>")
        .replace(/`([^`]+)`/g, "<code>$1</code>")
        .replace(/^- (.+)$/gm, "<li>$1</li>")
        .replace(/(<li>.*<\/li>\n?)+/g, "<ul>$&</ul>")
        .replace(/\n\n/g, "</p><p>")
        .replace(/^(?!<[huplo])/gm, "")
        .split("\n")
        .filter(Boolean)
        .join("\n");
}

const CSS = `
* { box-sizing: border-box; margin: 0; padding: 0; }
body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif; line-height: 1.6; color: #1a1a2e; background: #fafafa; max-width: 900px; margin: 0 auto; padding: 2rem 1rem; }
a { color: #3b82f6; text-decoration: none; }
a:hover { text-decoration: underline; }
h1 { font-size: 1.8rem; margin-bottom: 0.5rem; }
h2 { font-size: 1.3rem; margin: 1.5rem 0 0.5rem; border-bottom: 1px solid #e5e7eb; padding-bottom: 0.25rem; }
h3 { font-size: 1.1rem; margin: 1rem 0 0.25rem; }
p { margin: 0.5rem 0; }
code { background: #f1f5f9; padding: 0.15rem 0.35rem; border-radius: 3px; font-size: 0.9em; }
pre { background: #1e293b; color: #e2e8f0; padding: 1rem; border-radius: 6px; overflow-x: auto; margin: 1rem 0; }
pre code { background: none; color: inherit; padding: 0; }
ul { padding-left: 1.5rem; margin: 0.5rem 0; }
.badge { display: inline-block; padding: 0.15rem 0.5rem; border-radius: 9999px; font-size: 0.75rem; font-weight: 500; }
.type { background: #dbeafe; color: #1e40af; }
.tag { background: #fef3c7; color: #92400e; margin: 0 0.15rem; }
.chunk-card { border: 1px solid #e5e7eb; border-radius: 8px; padding: 1rem 1.25rem; margin-bottom: 0.75rem; background: white; transition: border-color 0.15s; }
.chunk-card:hover { border-color: #3b82f6; }
.chunk-card h3 { margin: 0 0 0.25rem; }
.chunk-card p { color: #6b7280; font-size: 0.875rem; margin: 0; }
.meta { color: #9ca3af; font-size: 0.8rem; margin-top: 0.5rem; }
.search { width: 100%; padding: 0.6rem 1rem; border: 1px solid #e5e7eb; border-radius: 6px; font-size: 1rem; margin-bottom: 1rem; }
.filters { display: flex; gap: 0.5rem; flex-wrap: wrap; margin-bottom: 1rem; }
.filter-btn { padding: 0.3rem 0.75rem; border: 1px solid #e5e7eb; border-radius: 9999px; background: white; cursor: pointer; font-size: 0.8rem; }
.filter-btn.active { background: #3b82f6; color: white; border-color: #3b82f6; }
.header { display: flex; justify-content: space-between; align-items: center; margin-bottom: 1.5rem; border-bottom: 1px solid #e5e7eb; padding-bottom: 1rem; }
.back { font-size: 0.9rem; }
.content { margin-top: 1rem; }
.rationale { background: #f0fdf4; border-left: 3px solid #22c55e; padding: 0.75rem 1rem; margin: 1rem 0; border-radius: 0 6px 6px 0; }
.nav { display: flex; gap: 1rem; align-items: center; }
`;

function generateIndex(chunks: any[]): string {
    const types = [...new Set(chunks.map((c: any) => c.type))];
    const typeButtons = types
        .map((t) => `<button class="filter-btn" onclick="filterByType('${t}')">${t}</button>`)
        .join("");

    const cards = chunks
        .map((c: any) => {
            const tags = (c.tags ?? [])
                .map((t: any) => {
                    const name = typeof t === "string" ? t : t.name;
                    return `<span class="badge tag">${escapeHtml(name)}</span>`;
                })
                .join(" ");
            const preview = escapeHtml((c.content ?? "").slice(0, 150)).replace(/\n/g, " ");
            return `<div class="chunk-card" data-type="${escapeHtml(c.type)}" data-title="${escapeHtml(c.title.toLowerCase())}">
    <h3><a href="chunk-${c.id}.html">${escapeHtml(c.title)}</a></h3>
    <p>${preview}${(c.content?.length ?? 0) > 150 ? "..." : ""}</p>
    <div class="meta"><span class="badge type">${escapeHtml(c.type)}</span> ${tags}</div>
</div>`;
        })
        .join("\n");

    return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>fubbik Knowledge Base</title><style>${CSS}</style></head>
<body>
<div class="header">
    <h1>fubbik Knowledge Base</h1>
    <div class="nav"><span class="meta">${chunks.length} chunks</span></div>
</div>
<input type="text" class="search" placeholder="Search chunks..." oninput="filterChunks(this.value)">
<div class="filters"><button class="filter-btn active" onclick="filterByType('')">All</button>${typeButtons}</div>
<div id="chunks">${cards}</div>
<script>
let activeType = '';
function filterChunks(query) {
    const q = query.toLowerCase();
    document.querySelectorAll('.chunk-card').forEach(el => {
        const matchesSearch = !q || el.dataset.title.includes(q) || el.textContent.toLowerCase().includes(q);
        const matchesType = !activeType || el.dataset.type === activeType;
        el.style.display = matchesSearch && matchesType ? '' : 'none';
    });
}
function filterByType(type) {
    activeType = type;
    document.querySelectorAll('.filter-btn').forEach(b => b.classList.remove('active'));
    event.target.classList.add('active');
    filterChunks(document.querySelector('.search').value);
}
</script>
</body></html>`;
}

function generateChunkPage(chunk: any): string {
    const tags = (chunk.tags ?? [])
        .map((t: any) => {
            const name = typeof t === "string" ? t : t.name;
            return `<span class="badge tag">${escapeHtml(name)}</span>`;
        })
        .join(" ");

    const content = markdownToHtml(chunk.content ?? "");

    let extras = "";
    if (chunk.rationale) {
        extras += `<div class="rationale"><strong>Rationale:</strong> ${escapeHtml(chunk.rationale)}</div>`;
    }
    if (chunk.alternatives?.length) {
        extras += `<h2>Alternatives Considered</h2><ul>${chunk.alternatives.map((a: string) => `<li>${escapeHtml(a)}</li>`).join("")}</ul>`;
    }
    if (chunk.consequences) {
        extras += `<h2>Consequences</h2><p>${escapeHtml(chunk.consequences)}</p>`;
    }

    return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>${escapeHtml(chunk.title)} -- fubbik</title><style>${CSS}</style></head>
<body>
<div class="header">
    <a href="index.html" class="back">&larr; Back to index</a>
    <div class="nav"><span class="badge type">${escapeHtml(chunk.type)}</span> ${tags}</div>
</div>
<h1>${escapeHtml(chunk.title)}</h1>
<div class="meta">Updated: ${new Date(chunk.updatedAt).toLocaleDateString()}</div>
<div class="content">${content}</div>
${extras}
</body></html>`;
}

export const exportSiteCommand = new Command("export-site")
    .description("Generate a static HTML site from the knowledge base")
    .option("-o, --output <dir>", "output directory", "fubbik-site")
    .option("--codebase <name>", "filter by codebase")
    .action(async (opts: { output: string; codebase?: string }, cmd: Command) => {
        const serverUrl = getServerUrl();
        if (!serverUrl) {
            outputError("Server URL required. Run 'fubbik init --server <url>'.");
            process.exit(1);
        }

        // Fetch all chunks
        const params = new URLSearchParams({ limit: "500", sort: "alpha" });
        if (opts.codebase) {
            const cbRes = await fetch(`${serverUrl}/api/codebases`);
            if (cbRes.ok) {
                const codebases = (await cbRes.json()) as { id: string; name: string }[];
                const match = codebases.find((c) => c.name === opts.codebase);
                if (match) params.set("codebaseId", match.id);
            }
        }

        const res = await fetch(`${serverUrl}/api/chunks?${params}`);
        if (!res.ok) {
            outputError(`Failed to fetch chunks: ${res.status}`);
            process.exit(1);
        }

        const data = (await res.json()) as { chunks?: any[] } | any[];
        const chunks = Array.isArray(data) ? data : (data.chunks ?? []);

        if (chunks.length === 0) {
            outputError("No chunks found.");
            process.exit(1);
        }

        // Create output directory
        mkdirSync(opts.output, { recursive: true });

        // Generate index
        writeFileSync(join(opts.output, "index.html"), generateIndex(chunks));

        // Generate individual pages
        for (const chunk of chunks) {
            writeFileSync(join(opts.output, `chunk-${chunk.id}.html`), generateChunkPage(chunk));
        }

        console.log(formatSuccess(`Generated ${chunks.length + 1} pages to ${opts.output}/`));
        console.log(`  Open: file://${join(process.cwd(), opts.output, "index.html")}`);

        output(cmd, { output: opts.output, pages: chunks.length + 1 }, "");
    });

File: crates/fubbik-cli/src/commands/export_site.rs
use std::collections::BTreeSet;
use std::path::Path;

use anyhow::{Context, Result, bail};

use crate::client::Client;
use crate::output::{self, OutputMode};

const CSS: &str = "*{box-sizing:border-box}body{font-family:system-ui,sans-serif;line-height:1.6;color:#172033;background:#fafafa;max-width:900px;margin:auto;padding:2rem 1rem}a{color:#2563eb;text-decoration:none}a:hover{text-decoration:underline}.header{display:flex;justify-content:space-between;align-items:center;border-bottom:1px solid #ddd;padding-bottom:1rem;margin-bottom:1.5rem}.search{width:100%;padding:.7rem;margin-bottom:1rem}.filters{display:flex;gap:.5rem;flex-wrap:wrap;margin-bottom:1rem}.filter-btn,.badge{border:1px solid #ddd;border-radius:999px;padding:.2rem .6rem;background:white}.chunk-card{border:1px solid #ddd;border-radius:8px;padding:1rem;margin-bottom:.75rem;background:white}.meta{color:#667085;font-size:.85rem}.content{margin-top:1rem}pre{background:#172033;color:#e2e8f0;padding:1rem;overflow:auto}.rationale{border-left:3px solid #22c55e;background:#f0fdf4;padding:.75rem 1rem}";

pub async fn run(
    client: &Client,
    output_dir: &Path,
    space: Option<&str>,
    mode: OutputMode,
) -> Result<()> {
    let space_id = client.resolve_space(space).await?;
    let chunks = super::export::load_chunks(client, space_id.as_deref()).await?;
    if chunks.is_empty() {
        bail!("no chunks found");
    }
    std::fs::create_dir_all(output_dir)
        .with_context(|| format!("could not create {}", output_dir.display()))?;
    std::fs::write(output_dir.join("index.html"), index_page(&chunks))?;
    for chunk in &chunks {
        let id = chunk["id"].as_str().context("export chunk omitted id")?;
        std::fs::write(
            output_dir.join(format!("chunk-{id}.html")),
            chunk_page(chunk),
        )?;
    }
    let result = serde_json::json!({"output": output_dir, "pages": chunks.len() + 1});
    match mode {
        OutputMode::Json => output::json(&result),
        OutputMode::Quiet => {
            println!("{}", output_dir.display());
            Ok(())
        }
        OutputMode::Human => {
            println!(
                "Generated {} pages to {}/",
                chunks.len() + 1,
                output_dir.display()
            );
            Ok(())
        }
    }
}

fn index_page(chunks: &[serde_json::Value]) -> String {
    let types = chunks
        .iter()
        .filter_map(|chunk| chunk["type"].as_str())
        .collect::<BTreeSet<_>>();
    let buttons = types
        .into_iter()
        .map(|kind| {
            format!(
                "<button class=\"filter-btn\" data-filter=\"{}\">{}</button>",
                escape(kind),
                escape(kind)
            )
        })
        .collect::<String>();
    let cards = chunks.iter().map(card).collect::<String>();
    format!(
        r#"<!doctype html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>fubbik Knowledge Base</title><style>{CSS}</style></head><body><div class="header"><h1>fubbik Knowledge Base</h1><span class="meta">{} chunks</span></div><input class="search" placeholder="Search chunks..."><div class="filters"><button class="filter-btn" data-filter="">All</button>{buttons}</div><div id="chunks">{cards}</div><script>let type='';const search=document.querySelector('.search');function apply(){{const q=search.value.toLowerCase();document.querySelectorAll('.chunk-card').forEach(c=>c.hidden=!((!type||c.dataset.type===type)&&(!q||c.textContent.toLowerCase().includes(q))))}}search.addEventListener('input',apply);document.querySelectorAll('.filter-btn').forEach(b=>b.addEventListener('click',()=>{{type=b.dataset.filter;apply()}}));</script></body></html>"#,
        chunks.len()
    )
}

fn card(chunk: &serde_json::Value) -> String {
    let content = chunk["content"].as_str().unwrap_or("");
    let preview = content.chars().take(150).collect::<String>();
    let suffix = if content.chars().count() > 150 {
        "..."
    } else {
        ""
    };
    format!(
        "<article class=\"chunk-card\" data-type=\"{}\"><h3><a href=\"chunk-{}.html\">{}</a></h3><p>{}{suffix}</p><div class=\"meta\">{}</div></article>",
        escape(chunk["type"].as_str().unwrap_or("note")),
        escape(chunk["id"].as_str().unwrap_or("")),
        escape(chunk["title"].as_str().unwrap_or("Untitled")),
        escape(&preview.replace('\n', " ")),
        tag_badges(chunk)
    )
}

fn chunk_page(chunk: &serde_json::Value) -> String {
    let title = escape(chunk["title"].as_str().unwrap_or("Untitled"));
    let mut extras = String::new();
    if let Some(rationale) = chunk["rationale"].as_str() {
        extras.push_str(&format!(
            "<div class=\"rationale\"><strong>Rationale:</strong> {}</div>",
            escape(rationale)
        ));
    }
    if let Some(alternatives) = chunk["alternatives"]
        .as_array()
        .filter(|items| !items.is_empty())
    {
        extras.push_str("<h2>Alternatives Considered</h2><ul>");
        for alternative in alternatives.iter().filter_map(serde_json::Value::as_str) {
            extras.push_str(&format!("<li>{}</li>", escape(alternative)));
        }
        extras.push_str("</ul>");
    }
    if let Some(consequences) = chunk["consequences"].as_str() {
        extras.push_str(&format!(
            "<h2>Consequences</h2><p>{}</p>",
            escape(consequences)
        ));
    }
    format!(
        "<!doctype html><html lang=\"en\"><head><meta charset=\"utf-8\"><meta name=\"viewport\" content=\"width=device-width,initial-scale=1\"><title>{title} — fubbik</title><style>{CSS}</style></head><body><div class=\"header\"><a href=\"index.html\">&larr; Back to index</a><div>{}</div></div><h1>{title}</h1><div class=\"meta\">Updated: {}</div><div class=\"content\">{}</div>{extras}</body></html>",
        tag_badges(chunk),
        escape(chunk["updatedAt"].as_str().unwrap_or("")),
        markdown_html(chunk["content"].as_str().unwrap_or(""))
    )
}

fn tag_badges(chunk: &serde_json::Value) -> String {
    chunk["tags"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|tag| tag.as_str().or_else(|| tag["name"].as_str()))
        .map(|tag| format!("<span class=\"badge\">{}</span> ", escape(tag)))
        .collect()
}

fn markdown_html(markdown: &str) -> String {
    markdown
        .lines()
        .map(|line| {
            let line = escape(line);
            if let Some(value) = line.strip_prefix("### ") {
                format!("<h3>{value}</h3>")
            } else if let Some(value) = line.strip_prefix("## ") {
                format!("<h2>{value}</h2>")
            } else if let Some(value) = line.strip_prefix("# ") {
                format!("<h1>{value}</h1>")
            } else if let Some(value) = line.strip_prefix("- ") {
                format!("<li>{value}</li>")
            } else if line.is_empty() {
                String::new()
            } else {
                format!("<p>{line}</p>")
            }
        })
        .collect()
}

fn escape(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#39;")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generated_pages_escape_chunk_content_and_include_metadata() {
        // Given a chunk containing HTML-like user content and decision metadata
        let chunk = serde_json::json!({"id":"c1","title":"<Title>","type":"decision","content":"# Safe\n<script>x</script>","updatedAt":"2026-01-01","tags":[{"name":"security"}],"rationale":"Because","alternatives":["Other"],"consequences":"Safer"});
        // When index and detail pages are generated
        let index = index_page(std::slice::from_ref(&chunk));
        let detail = chunk_page(&chunk);
        // Then navigation metadata remains and raw HTML is escaped
        assert!(index.contains("chunk-c1.html"));
        assert!(detail.contains("&lt;script&gt;x&lt;/script&gt;"));
        assert!(detail.contains("Rationale:"));
        assert!(!detail.contains("<script>x</script>"));
    }
}
